File: anova_analysis.py
"""
Phân tích ANOVA - So sánh giá khách sạn theo hạng sao
"""

import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.formula.api import ols
from scipy import stats
from scipy.stats import studentized_range
import sys
import io

# Thiết lập encoding UTF-8 cho output
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')


def main():
    # --- Đọc file ---
    file_path = "khach_san.xlsx"  # Thay đổi đường dẫn nếu cần
    
    try:
        sheet1 = pd.read_excel(file_path, sheet_name="giá_khách_sạn_theo_ngày")
        sheet2 = pd.read_excel(file_path, sheet_name="Thông tin khách sạn")
    except FileNotFoundError:
        print(f"❌ Không tìm thấy file: {file_path}")
        print("Vui lòng đặt file 'khach_san.xlsx' trong cùng thư mục hoặc cập nhật đường dẫn.")
        return
    
    # --- Làm sạch tên cột ---
    sheet1.columns = sheet1.columns.str.strip()
    sheet2.columns = sheet2.columns.str.strip()
    
    # --- Làm sạch cột giá ---
    sheet1["price_on_list"] = (
        sheet1["price_on_list"]
        .astype(str)
        .str.replace("VND", "", regex=False)
        .str.replace(".", "", regex=False)
        .str.strip()
    )
    
    sheet1["price_on_list"] = pd.to_numeric(sheet1["price_on_list"], errors="coerce")
    
    # --- Gộp dữ liệu ---
    data = pd.merge(sheet1, sheet2, on="hotel_id", how="inner")
    
    # --- Lọc dữ liệu hợp lệ ---
    data = data[["hotel_id", "price_on_list", "star"]].dropna()
    data = data[data["price_on_list"] > 0]
    data["star"] = data["star"].astype(int)
    
    # --- Kiểm tra nhóm sao ---
    print("Nhóm sao:", data["star"].unique())
    
    # ===== BẢNG SỐ LƯỢNG MỖI NHÓM =====
    group_count = data.groupby("star")["price_on_list"].count().reset_index()
    group_count.columns = ["Số sao", "Số lượng khách sạn"]
    
    print("\n===== BẢNG SỐ LƯỢNG MỖI NHÓM =====")
    print(group_count.to_markdown(index=False, tablefmt="grid", numalign="left", stralign="left"))
    
    # --- ANOVA ---
    model = ols("price_on_list ~ C(star)", data=data).fit()
    anova_table = sm.stats.anova_lm(model, typ=2)
    
    # --- Thêm cột Mean Square (MS) ---
    anova_table["MS"] = anova_table["sum_sq"] / anova_table["df"]
    
    # --- F critical ---
    df_between = anova_table["df"].iloc[0]
    df_within = anova_table["df"].iloc[1]
    f_critical = stats.f.ppf(0.95, df_between, df_within)
    
    anova_display = anova_table.rename(columns={
        "sum_sq": "SS",
        "df": "df",
        "F": "F",
        "PR(>F)": "P-value"
    })[["SS", "df", "MS", "F", "P-value"]]
    
    print("\n===== BẢNG KẾT QUẢ ANOVA =====")
    print(anova_display.to_markdown(tablefmt="grid", numalign="left", stralign="left"))
    print(f"\nF critical (α = 0.05): {f_critical:.4f}")
    
    p_value = anova_table["PR(>F)"].iloc[0]
    
    # --- Kết luận ANOVA ---
    if p_value < 0.05:
        print("\n✅ Có sự khác biệt có ý nghĩa thống kê giữa các nhóm sao (p < 0.05)")
        
        # --- KIỂM ĐỊNH HSD THỦ CÔNG ---
        print("\n===== KIỂM ĐỊNH HSD THỦ CÔNG =====")
        
        MSE = anova_table["MS"].iloc[1]  # Mean Square Error
        df_within = int(anova_table["df"].iloc[1])
        k = data["star"].nunique()
        n_per_group = data.groupby("star")["price_on_list"].count().mean()
        q_critical = studentized_range.ppf(0.95, k, df_within)
        
        HSD = q_critical * np.sqrt(MSE / n_per_group)
        print(f"HSD (ngưỡng chênh lệch cần đạt) = {HSD:,.2f}")
        
        # --- Trung bình từng nhóm ---
        means = data.groupby("star")["price_on_list"].mean().sort_index()
        means_df = means.reset_index()
        means_df.columns = ["star", "price_on_list"]
        means_df["price_on_list"] = means_df["price_on_list"].apply(lambda x: f"{x:,.0f}")
        
        print("\nTrung bình giá theo sao:")
        print(means_df.to_markdown(index=False, tablefmt="grid", numalign="left", stralign="left"))
        
        # --- So sánh từng cặp ---
        pairs = []
        for i in range(len(means)):
            for j in range(i + 1, len(means)):
                diff = abs(means.iloc[i] - means.iloc[j])
                significant = diff > HSD
                pairs.append({
                    "group1": means.index[i],
                    "group2": means.index[j],
                    "mean_diff": diff,
                    "Significant": "✅ Có khác biệt" if significant else "❌ Không khác biệt"
                })
        
        pairs_df = pd.DataFrame(pairs)
        
        # --- In bảng đẹp (căn trái) ---
        pairs_df_display = pairs_df.copy()
        pairs_df_display["mean_diff"] = pairs_df_display["mean_diff"].apply(lambda x: f"{x:,.0f}")
        print("\n===== SO SÁNH THỦ CÔNG THEO HSD =====")
        print(pairs_df_display.to_markdown(index=False, tablefmt="grid", numalign="left", stralign="left"))
        
        # --- Kết luận cuối ---
        sig = pairs_df[pairs_df["Significant"] == "✅ Có khác biệt"]
        if not sig.empty:
            print("\n✅ Các cặp nhóm có khác biệt theo HSD:")
            for _, row in sig.iterrows():
                print(f"   - Nhóm {row['group1']} và nhóm {row['group2']} khác biệt (Δ = {row['mean_diff']:.2f})")
        else:
            print("\n❌ Không có cặp nhóm nào khác biệt theo HSD.")
    
    else:
        print("\n❌ Không có sự khác biệt đáng kể giữa các nhóm sao (p ≥ 0.05)")


if __name__ == "__main__":
    main()

File: anova_analysis.R
# Phan tich ANOVA - So sanh gia khach san theo hang sao

# Thiet lap encoding UTF-8
options(encoding = "UTF-8")

# --- Cai dat va load thu vien ---
# install.packages(c("readxl", "dplyr", "tidyr", "knitr"))

library(readxl)
library(dplyr)
library(tidyr)
library(knitr)

# --- Ham dinh dang so ---
format_number <- function(x) {
  format(x, big.mark = ",", scientific = FALSE, trim = TRUE)
}

# --- Doc file ---
# Thiet lap thu muc lam viec
setwd("E:/Kha/UIT/hk1_2025")

file_path <- "khach_san.xlsx"  # Thay doi duong dan neu can

# Kiem tra file ton tai
if (!file.exists(file_path)) {
  cat("Thu muc hien tai:", getwd(), "\n")
  cat("Cac file co san:", paste(list.files(), collapse = ", "), "\n")
  stop("Khong tim thay file: ", file_path, 
       "\nVui long dat file 'khach_san.xlsx' trong thu muc lam viec hien tai.")
}

# Doc du lieu - dung sheet index thay vi ten sheet
sheet1 <- read_excel(file_path, sheet = 1)
sheet2 <- read_excel(file_path, sheet = 2)

# --- Lam sach ten cot ---
names(sheet1) <- trimws(names(sheet1))
names(sheet2) <- trimws(names(sheet2))

# --- Lam sach cot gia ---
sheet1 <- sheet1 %>%
  mutate(
    price_on_list = gsub("VND", "", price_on_list, fixed = TRUE),
    price_on_list = gsub("\\.", "", price_on_list),
    price_on_list = gsub("\u00A0", "", price_on_list, fixed = TRUE),  # Remove non-breaking space
    price_on_list = as.numeric(trimws(price_on_list))
  )

# --- Gop du lieu ---
data <- inner_join(sheet1, sheet2, by = "hotel_id")

# --- Loc du lieu hop le ---
data <- data %>%
  select(hotel_id, price_on_list, star) %>%
  drop_na() %>%
  filter(price_on_list > 0) %>%
  mutate(star = as.integer(star))

# --- Kiem tra nhom sao ---
cat("Nhom sao:", unique(data$star), "\n")

# ===== BANG SO LUONG MOI NHOM =====
group_count <- data %>%
  group_by(star) %>%
  summarise(So_luong_khach_san = n()) %>%
  rename(So_sao = star)

cat("\n===== BANG SO LUONG MOI NHOM =====\n")
print(kable(group_count, format = "pipe"))

# --- ANOVA ---
# Chuyen star thanh factor de ANOVA
data$star <- as.factor(data$star)

# Thuc hien ANOVA
anova_model <- aov(price_on_list ~ star, data = data)
anova_summary <- summary(anova_model)

# Lay bang ANOVA
anova_table <- anova_summary[[1]]

# Tinh toan cac gia tri
SS_between <- anova_table$"Sum Sq"[1]
SS_within <- anova_table$"Sum Sq"[2]
df_between <- anova_table$Df[1]
df_within <- anova_table$Df[2]
MS_between <- anova_table$"Mean Sq"[1]
MS_within <- anova_table$"Mean Sq"[2]
F_value <- anova_table$"F value"[1]
p_value <- anova_table$"Pr(>F)"[1]

# F critical
f_critical <- qf(0.95, df_between, df_within)

# Tao bang hien thi
anova_display <- data.frame(
  Source = c("C(star)", "Residual"),
  SS = c(format_number(SS_between), format_number(SS_within)),
  df = c(df_between, df_within),
  MS = c(format_number(MS_between), format_number(MS_within)),
  F = c(sprintf("%.4f", F_value), ""),
  P_value = c(sprintf("%.6e", p_value), ""),
  check.names = FALSE
)

cat("\n===== BANG KET QUA ANOVA =====\n")
print(kable(anova_display, format = "pipe", align = "l"))
cat(sprintf("\nF critical (alpha = 0.05): %.4f\n", f_critical))

# --- Ket luan ANOVA ---
if (p_value < 0.05) {
  cat("\n>> Co su khac biet co y nghia thong ke giua cac nhom sao (p < 0.05)\n")
  
  # --- KIEM DINH HSD THU CONG ---
  cat("\n===== KIEM DINH HSD THU CONG =====\n")
  
  # Tinh HSD
  MSE <- MS_within
  k <- length(unique(data$star))
  n_per_group <- mean(table(data$star))
  
  # q critical tu phan phoi studentized range
  q_critical <- qtukey(0.95, k, df_within)
  
  HSD <- q_critical * sqrt(MSE / n_per_group)
  cat(sprintf("HSD (nguong chenh lech can dat) = %s\n", format_number(round(HSD, 2))))
  
  # --- Trung binh tung nhom ---
  means <- data %>%
    group_by(star) %>%
    summarise(mean_price = mean(price_on_list)) %>%
    arrange(star)
  
  means_display <- means %>%
    mutate(price_on_list = format_number(round(mean_price, 0))) %>%
    select(star, price_on_list)
  
  cat("\nTrung binh gia theo sao:\n")
  print(kable(means_display, format = "pipe", align = "l"))
  
  # --- So sanh tung cap ---
  pairs <- data.frame()
  
  for (i in 1:(nrow(means) - 1)) {
    for (j in (i + 1):nrow(means)) {
      diff <- abs(means$mean_price[i] - means$mean_price[j])
      significant <- diff > HSD
      
      pairs <- rbind(pairs, data.frame(
        group1 = as.character(means$star[i]),
        group2 = as.character(means$star[j]),
        mean_diff = diff,
        Significant = ifelse(significant, "Co khac biet", "Khong khac biet")
      ))
    }
  }
  
  # --- In bang dep ---
  pairs_display <- pairs %>%
    mutate(mean_diff = format_number(round(mean_diff, 0)))
  
  cat("\n===== SO SANH THU CONG THEO HSD =====\n")
  print(kable(pairs_display, format = "pipe", align = "l", row.names = FALSE))
  
  # --- Ket luan cuoi ---
  sig_pairs <- pairs %>% filter(Significant == "Co khac biet")
  
  if (nrow(sig_pairs) > 0) {
    cat("\n>> Cac cap nhom co khac biet theo HSD:\n")
    for (i in 1:nrow(sig_pairs)) {
      cat(sprintf("   - Nhom %s va nhom %s khac biet (Delta = %.2f)\n",
                  sig_pairs$group1[i], sig_pairs$group2[i], sig_pairs$mean_diff[i]))
    }
  } else {
    cat("\nKhong co cap nhom nao khac biet theo HSD.\n")
  }
  
} else {
  cat("\nKhong co su khac biet dang ke giua cac nhom sao (p >= 0.05)\n")
}

cat("\n=== Hoan tat phan tich ===\n")
